libs/cardano-ledger-core/src/Cardano/Ledger/Core/PParams.hs

```diff
@@ -546,3 +559,46 @@ downgradePParamsUpdate ::
   PParamsUpdate (PreviousEra era)
 downgradePParamsUpdate args (PParamsUpdate pphkd) =
   PParamsUpdate (downgradePParamsHKD @_ @StrictMaybe args pphkd)
+
+-- =====================================================================================
+-- Tools for building ToPlutusData instances for (PParamUpdates era).
+
+-- | Pair the tag, and exisitenially hide the type of the lens for the field with that Lens'
+data PParam era where
+  PParam :: ToPlutusData t => Word -> Lens' (PParamsUpdate era) (StrictMaybe t) -> PParam era
+
+-- | Turn a list into a Map, this assures we have no duplicates.
+makePParamMap :: [PParam era] -> Map Word (PParam era)
+makePParamMap xs = Map.fromList [(n, p) | p@(PParam n _) <- xs]
+
+-- ====================================================================
+-- Generic, Table (Map Word (PParam era)) driven translators for
+-- PParamUpdate, A table for some era, will make instances for (PParamUpdate era) for that Era.
+
+-- | Translate a PParamUpdate to Data, using the PParamMap.
+-- The PParamUpdate is Basically (Map (tag t) (Some (StrictMaybe t)))
+-- The idea is to only store (Map (tag t) (Some t)) for just the components
+-- that had (SJust t), and to leave out the other tags. To transform this
+-- back to a (Map (tag t) (Some t)), we start with a full map where everything is
+-- SNothing, and then override the tags we find in the data with (SJust t)
+-- That is what happens in pparamUpdateFromData below
+pparamUpdateToData :: Map Word (PParam era) -> PParamsUpdate era -> Data
+pparamUpdateToData pparamMap ppu = Map (Map.foldrWithKey' toMapPair [] pparamMap)
+  where
+    toMapPair tag (PParam _ ppuL) ans =
+      case ppu ^. ppuL of
+        SNothing -> ans
+        SJust t -> (I (toInteger @Word tag), toPlutusData t) : ans
+
+-- | Translate Data to a PParamUpdate, using the PParamMap.
+pparamUpdateFromData :: EraPParams era => Map Word (PParam era) -> Data -> Maybe (PParamsUpdate era)
+pparamUpdateFromData pparamMap (Map pairs) =
+  fst
+    <$> foldlM accum (emptyPParamsUpdate, pparamMap) pairs
+  where
+    accum (!ppu, !leftOverMap) (key, value) = do
+      tag <- fromPlutusData key
+      PParam _ ppuL <- Map.lookup tag leftOverMap
+      val <- fromPlutusData value
+      pure (ppu & ppuL .~ SJust val, Map.delete tag leftOverMap)
+pparamUpdateFromData _ _ = Nothing
```

Comment:
It would be better to put these two functions into a new module `Cardano.Ledger.Conway.Plutus.Context` together with `ConwayEraPlutusTxInfo`, because those functions will be used for creating instances for that class. This way it will be easier for people to find.

This functionality is very much plutus specific and can't be used prior to Conway, so we don't need to to keep it in the `Cardano.Ledger.Core.PParams`

The `PParam` GADT, on the other hand, we'll be able to use other places later on, so this is a perfect home for it.

Reply:
done